File: calc2/src/main/java/jp/co/yoshida/katsushige/calc2/Point3D.java
package jp.co.yoshida.katsushige.calc2;


/**
 * 三次元座標クラス
 * javafx.geometry のPoint3D互換の予定
 */
public class Point3D {
    public double X;
    public double Y;
    public double Z;

    /**
     * コンストラクタ Point3Dの新しいインスタンスを作成
     */
    public Point3D() {

    }

    /**
     * コンストラクタ Point3Dの新しいインスタンスを作成
     * @param x
     * @param y
     * @param z
     */
    public Point3D(double x, double y, double z) {
        X = x;
        Y = y;
        Z = z;
    }

    /**
     * コンストラクタ Point3Dの新しいインスタンスを作成
     * @param point
     */
    public Point3D(Point3D point) {
        X = point.X;
        Y = point.Y;
        Z = point.Z;
    }

    /**
     * この点の座標に指定された座標が加算された点を返します。
     * @param x
     * @param y
     * @param z
     * @return
     */
    public Point3D add(double x, double y, double z) {
        X += x;
        Y += y;
        Z += z;
        return new Point3D(X, Y, Z);
    }

    /**
     * この点の座標に指定された座標が加算された点を返します。
     * @param p
     * @return
     */
    public Point3D add(Point3D p) {
        X += p.X;
        Y += p.Y;
        Z += p.Z;
        return new Point3D(X, Y, Z);
    }

    /**
     * この点で表されるベクトルと指定されたベクトルの間の角度(rad)を計算します。
     * @param x
     * @param y
     * @param z
     * @return
     */
    public double angle(double x, double y, double z) {
        double c = (X*x+Y*y+Z*z) /
                (Math.sqrt(X*X+Y*Y+Z*Z)*Math.sqrt(x*x+y*y+z*z));
        return Math.acos(c);
    }

    /**
     * この点で表されるベクトルと指定されたベクトルの間の角度(rad)を計算します。
     * @param point
     * @return
     */
    public double angle(Point3D point) {
        double c = (X*point.X+Y*point.Y+Z*point.Z) /
                (Math.sqrt(X*X+Y*Y+Z*Z)*Math.sqrt(point.X*point.X+point.Y*point.Y+point.Z*point.Z));
        return Math.acos(c);
    }

    /**
     * このインスタンスで表されるベクトルと指定されたベクトルのクロス積(外積)を計算します。
     * @param x
     * @param y
     * @param z
     * @return
     */
    public Point3D crossProduct(double x, double y, double z) {
        Point3D vec = new Point3D();
        vec.X = Y*z - Z*y;
        vec.Y = Z*x - X*z;
        vec.Z = X*y - Y*x;
        return  vec;
    }

    /**
     * このインスタンスで表されるベクトルと指定されたベクトルのクロス積(外積)を計算します。
     * @param point
     * @return
     */
    public Point3D crossProduct(Point3D point) {
        Point3D vec = new Point3D();
        vec.X = Y*point.Z - Z*point.Y;
        vec.Y = Z*point.X - X*point.Z;
        vec.Z = X*point.Y - Y*point.X;
        return  vec;
    }

    /**
     * この点と指定されたpointの間の距離を計算します。
     * @param x
     * @param y
     * @param z
     * @return
     */
    public double distance(double x, double y, double z) {
        return  Math.sqrt((X-x)*(X-x)+(Y-y)*(Y-y)+(Z-z)*(Z-z));
    }

    /**
     * この点と指定されたpointの間の距離を計算します。
     * @param point
     * @return
     */
    public double distance(Point3D point) {
        return  Math.sqrt((X-point.X)*(X-point.X)+(Y-point.Y)*(Y-point.Y)+(Z-point.Z)*(Z-point.Z));
    }
    public double getX() {
        return X;
    }
    public double getY() {
        return Y;
    }
    public double getZ() {
        return Z;
    }
    public Point3D max(Point3D point) {
        Point3D max = new Point3D();
        max.X = Math.max(X,point.X);
        max.Y = Math.max(Y,point.X);
        max.Z = Math.max(Z,point.Z);
        return max;
    }
    public Point3D min(Point3D point) {
        Point3D min = new Point3D();
        min.X = Math.min(X,point.X);
        min.Y = Math.min(Y,point.X);
        min.Z = Math.min(Z,point.Z);
        return min;
    }

    public String toString() {
        String buf = X+","+Y+","+Z;
        return buf;
    }
}

File: mylib/src/main/java/jp/co/yoshida/katsushige/mylib/Point3D.kt
package jp.co.yoshida.katsushige.mylib

import kotlin.math.acos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

class Point3D() {

    var x = 0.0
    var y = 0.0
    var z = 0.0

    /**
     * Constructors
     */
    constructor(x: Int, y: Int, z: Int): this() {
        this.x = x.toDouble()
        this.y = y.toDouble()
        this.z = z.toDouble()
    }

    /**
     * Constructors
     */
    constructor(x: Double, y: Double, z:Double): this() {
        this.x = x
        this.y = y
        this.z = z
    }

    constructor(p: Point3D): this() {
        this.x = p.x
        this.y = p.y
        this.z = p.z
    }

    override fun toString(): String {
        return x.toString() + "," + y.toString() + "," + z.toString()
    }

    /**
     * XY座標を2次元変換
     * return   2D XY座標
     */
    fun toPointXY() : PointD {
        return PointD(x, y)
    }

    /**
     * XY座標を2次元変換
     * return   2D YZ座標
     */
    fun toPointYZ() : PointD {
        return PointD(y, z)
    }

    /**
     * XY座標を2次元変換
     * return   2D ZX座標
     */
    fun toPointZX() : PointD {
        return PointD(z, x)
    }

    /**
     * 座標を0クリアする
     */
    fun clear() {
        x = 0.0
        y = 0.0
        z = 0.0
    }

    /**
     * 座標値が0かを確認する
     */
    fun isEmpty(): Boolean {
        if (x == 0.0 && y == 0.0 && z == 0.0)
            return true
        else
            return false
    }

    /**
     * 長さ、原点からの距離
     * return       長さ
     */
    fun length(): Double {
        return sqrt(x * x + y * y + z * z)
    }

    /**
     * 座標データの符号を反転する
     */
    fun inverse() {
        this.x *= -1.0
        this.y *= -1.0
        this.z *= -1.0
    }

    /**
     * 座標をオフセット分移動させる
     */
    fun offset(offset: Point3D) {
        this.x += offset.x;
        this.y += offset.y;
        this.z += offset.z;
    }

    /**
     * この点で表されるベクトルと指定されたベクトルの間の角度(rad)を計算
     * 原点を中心とした2点の角度(内積から角度を求める)
     * p            対象ベクトル
     * return       角度(rad)
     */
    fun angle(p: Point3D): Double {
        val c = (x * p.x + y * p.y + z * p.z) /
                (sqrt(x * x + y * y + z * z) * sqrt(p.x * p.x + p.y * p.y + p.z * p.z))
        return acos(c)
    }

    /**
     * 原点を起点とした2つのベクトルの外積
     * p            対象ベクトル
     * return       外積
      */
    fun crossProduct(p: Point3D): Point3D {
        return Point3D(y * p.z - z * p.y, z * p.x - x * p.z, x * p.y - y * p.z)
    }

    /**
     * 指定点との距離
     * return       距離
     */
    fun distance(p: Point3D): Double{
        val dx = x - p.x
        val dy = y - p.y
        val dz = z - p.z
        return sqrt(dx * dx + dy * dy + dz *dz)
    }

    /**
     * 比較して大きい値を返す
     * p        比較座標
     * return   比較値
      */
    fun max(p: Point3D): Point3D {
        return Point3D(max(x, p.x), max(y, p.y), max(z, p.z))
    }

    /**
     * 比較して大きい値を返す
     * p        比較座標
     * return   比較値
     */
    fun min(p: Point3D): Point3D {
        return Point3D(min(x, p.x), min(y, p.y), min(z, p.z))
    }

}